unicodetools/src/test/java/org/unicode/text/UCD/TestCodeInvariants.java

```diff
@@ -230,4 +242,132 @@ private static String showInfo(
+                                + property
+                                + ")");
+            }
+        }
+        Set<Object> unicodeSetUnaryQueryNames =
+                Arrays.stream(UcdProperty.values())
+                        .filter(p -> p.getType() == PropertyType.Binary)
+                        .collect(Collectors.toCollection(() -> new HashSet<>()));
+        unicodeSetUnaryQueryNames.addAll(
+                Arrays.asList(UcdPropertyValues.General_Category_Values.values()));
+        unicodeSetUnaryQueryNames.addAll(Arrays.asList(UcdPropertyValues.Script_Values.values()));
+        testLM3NamespaceUniqueness(
+                unicodeSetUnaryQueryNames,
+                x ->
+                        x instanceof UcdProperty
+                                ? ((UcdProperty) x).getNames().getAllNames()
+                                : ((Named) x).getNames().getAllNames(),
+                Set.of("Age"),
+                "Violation of UnicodeSet requirements: gc-sc-binary property namespace collision");
+        Set<Object> nonCollidingProperties = new HashSet<>();
+        nonCollidingProperties.addAll(Arrays.asList(UcdProperty.values()));
+        nonCollidingProperties.addAll(
+                Arrays.asList(UcdPropertyValues.General_Category_Values.values()));
+        nonCollidingProperties.addAll(Arrays.asList(UcdPropertyValues.Script_Values.values()));
+        nonCollidingProperties.remove(UcdProperty.ISO_Comment); // Collides with gc Other.
+        nonCollidingProperties.remove(UcdProperty.Case_Folding); // Collides with gc Format.
+        nonCollidingProperties.remove(
+                UcdProperty.Lowercase_Mapping); // Collides with gc Cased_Letter.
+        nonCollidingProperties.remove(UcdProperty.Script); // Collides with gc Currency_Symbol.
+        testLM3NamespaceUniqueness(
+                nonCollidingProperties,
+                x ->
+                        x instanceof UcdProperty
+                                ? ((UcdProperty) x).getNames().getAllNames()
+                                : ((Named) x).getNames().getAllNames(),
+                Set.of("Age"),
+                "Unusual (not a violation of UnicodeSet requirement): New gc-sc-non-binary property namespace collision");
+    }
+
+    <T> void testLM3NamespaceUniqueness(
+            Iterable<T> namespace,
+            Function<T, List<String>> getNames,
+            Set<String> expectedRedundant,
+            String message) {
+        final Map<String, T> entitiesByAlias = new TreeMap<>(UnicodeProperty.PROPERTY_COMPARATOR);
+        final Map<String, String> aliasesByLM3Skeleton = new HashMap<>();
+        for (T entity : namespace) {
+            for (String alias : getNames.apply(entity)) {
+                final var matchingEntity = entitiesByAlias.get(alias);
+                final var lm3Skeleton = UnicodeProperty.toSkeleton(alias);
+                final var matchingAlias = aliasesByLM3Skeleton.get(lm3Skeleton);
```

Comment:
random thought:
if the comparator was mutable and could remember the values when they compared equal, then you wouldn't need this extra Map because you could fetch the earlier alias from the comparator state...

Reply:
Yes, but squirreling things away in mutable state feels weird.
If getEntry on the TreeMap had been public rather than an implementation detail of get, I could have use that, but alas…